plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py

```diff
@@ -309,68 +310,55 @@ def remove_unpinned_files(self) -> None:
     def download(
         self,
         hash_id: str,
-        target_dir: str,
+        target_dir: Union[str, Path],
         fix_path: bool = True,
         attempts: int = 5,
     ) -> str:
         """
         Download dir by its hash.
 
-        :param hash_id: str. hash of file to download
-        :param target_dir: str. directory to place downloaded
+        :param hash_id: str. hash of file or package to download
+        :param target_dir: Union[str, Path]. directory to place downloaded
         :param fix_path: bool. default True. on download don't wrap result in to hash_id directory.
         :param attempts: int. default 5. How often to attempt the download.
         :return: downloaded path
         """
 
-        if not os.path.exists(target_dir):  # pragma: nocover
-            os.makedirs(target_dir, exist_ok=True)
+        def move_to_target_dir(download_path) -> str:
+            """Move downloaded content to target directory"""
+
+            if download_path.is_file():
+                shutil.copy(download_path, target_dir)
+                return str(target_dir)
+
+            # else it is a directory containing a single package path
+            package_path = download_path
+            if fix_path:
+                # assumption is it contains one nested directory: the package
+                paths = list(download_path.glob("*"))
+                assumption_is_valid = len(paths) == 1 and paths[0].is_dir()
+                if not assumption_is_valid:  # pragma: no cover
+                    error_msg = f"Expected a single directory, found: {paths}"
+                    raise DownloadError(error_msg)
+                package_path = paths.pop()
```

Comment:
added explicit check for the apparent assumption. Relates to this original piece of code
https://github.com/valory-xyz/open-aea/blob/67ac9a1c6545b1a675e078c959059ff330d2805d/plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py#L350-L358

I am of the opinion this should not exist here and be moved to the one place the returned path is actually used, but that is outside of the scope of this PR
https://github.com/valory-xyz/open-aea/blob/beadc9f172e4c331e4e95e73d17bfb4ddbd60d38/aea/cli/fetch.py#L147-L155

Reply:
nice refactor.

but i dont like the mix of functionalities in download method.

i think we need quite simple download with retry method
ans some high level method that uses simple download and applies all the package related things

`download` method should know nothing about packages, files and how handle data downloaded.

Reply:
100% agree (as per #501), it's mostly that I try to limit the amount of changes in a single PR. Now with the nested function it should be easier to achieve in a future refactor.